Python 3 support

pallegro opened the pull request:
A modest step toward python 3 support in ROS.  Fixes all the problems detailed in the [wiki](http://wiki.ros.org/python_2_and_3_compatible_code) as well as the tf extension module.  Most are simply string replacements, but based on python 3 docs and [_roslz4module.c](https://github.com/ros/ros_comm/blob/indigo-devel/utilities/roslz4/src/_roslz4module.c), in pytf.cpp all module state is moved from global variables to a struct linked to each initialization.   This requires a bit of code shuffling and isn't strictly necessary.   I have another [version](https://github.com/pallegro/geometry/blob/indigo-devel/tf/src/pytf.cpp) which leaves things as they are.
Tested with python 2.7 and 3.4.  Note that tf/test/testPython.py has two failing test cases with or without these changes (getLatestCommonTime with no common ancestor returns time(0) rather than raising an exception).

pallegro committed:
Python 3 fixes

pallegro committed:
Python 3 fixes

pallegro committed:
Python 3 fixes

pallegro committed:
Python 3 fixes for tf extension module (no more global state version)

ros-pull-request-builder commented:
Can one of the admins verify this patch?

de-vri-es commented:
Can we get an update on this PR? Python 3 support should be a priority imo. It would be extremely nice to have python3 support in Kinetic.

I was working on this myself but this PR is more complete than mine. Although it has conflicts now.

tfoote commented:
We were pushing hard on python3 support for a while, but it lost steam when we realized that there was no effective way to change the default python implementation for packaging.

I'd be willing to merge the patch if it can be verified. But if I was to work on an implementation I'd much rather work on porting [tf2_py](https://github.com/ros/geometry2/tree/indigo-devel/tf2_py)/[tf2_ros](https://github.com/ros/geometry2/tree/indigo-devel/tf2_ros) This package is they python bindings for the backwards compatibility package tf which has been replaced by tf2.

de-vri-es commented:
I'm building ROS Kinetic packages for Arch Linux with python 3 as the default interpreter. However, `tf` failed to build. The same goes for `tf2`, but I also put in a pull request for `tf2_py` (ros/geometry2#173).

I did more or less the same for `tf` here: [delftrobotics/geometry/tree/pytf-python3](https://github.com/delftrobotics/geometry/tree/pytf-python3). But then I noticed this PR which does more than what I did. I only fixed the compile errors in `pytf.cpp` and the print calls in `tfwtf.py`.

Currently I'm building from my fork, but I really would like to see python3 fully supported in the official repositories. And for now it doesn't look like dropping `tf` for `tf2` is a viable option. There are plenty of packages still depending on `tf`.

But to get back to the matter at hand, this PR doesn't merge currently, so what is the way forward? @pallegro Do you want to fix the merge conflicts, or shall I give it a go? Alternatively I could continue on my fork, but that seems like a waste of this PR.

pallegro commented:
I rebased both versions of this PR.  Sadly, my git-fu is weak, so since I rebased the other branch first I haven't been able to get this one in order. For now I created a new branch: [py3_module_reinit_rebased](https://github.com/pallegro/geometry/tree/py3_module_reinit_rebased)
@de-vri-es can you test this?  I will as well, but I don't have a python3 ros core going right now and it will be a few days until I can get to this. I'm also on Arch so if you want any help testing other py3 stacks let me know.

de-vri-es commented:
Thanks for rebasing :)

I'll test your branch hopefully this monday. I'm not 100% sure I'll have the time though. If not I should have time at the latest next weekend

de-vri-es commented:
Now that #151 is merged, this can be closed I think?

tfoote commented:
I believe so too. Thanks!
